### PR TITLE
Savegame

### PR DESCRIPTION
**Pull request by theis999**
Adds the ability to save the game

### CODE REVIEW

File: cMain.cpp [hunk 1]
```diff
@@ -2918,7 +2941,7 @@ void cMain::extract_parameters() {
 	task = extract_task();
 	x_cord = extract_x_cord();
 	y_cord = extract_y_cord();
-	units = extract_units();
+	units = task == "Save" ? extract_units(false) : extract_units();
```
Comment (MortenTobiasNielsen) [inline cMain.cpp:2944]: I don't expect to use units for multiple things which isn't a number, so I rather not have this distinction and handle it in the extract_units method.

File: cMain.cpp [hunk 2]
```diff
@@ -3040,7 +3067,8 @@ std::string cMain::extract_y_cord() {
 	return std::to_string(wxAtof(txt_y_cord->GetValue()));
 }
 
-std::string cMain::extract_units() {
+std::string cMain::extract_units(bool validate) {
+	if (!validate) return txt_units->GetValue().ToStdString();
```
Comment (MortenTobiasNielsen) [inline cMain.cpp:3071]:
```
std::string cMain::extract_units() {
	if (rbtn_save->GetValue()) {
		return txt_units->GetValue().ToStdString();
	}
```

File: cMain.h [hunk 3]
```diff
@@ -318,7 +320,7 @@ class cMain : public GUI_Base {
 	std::string extract_task();
 	std::string extract_x_cord();
 	std::string extract_y_cord();
-	std::string extract_units();
+	std::string extract_units(bool validate = true);
```
Comment (MortenTobiasNielsen) [inline cMain.h:323]: I don't expect to use units for multiple things which isn't a number, so I rather not have this distinction and handle it in the extract_units method.

**Review by MortenTobiasNielsen — CHANGES_REQUESTED** [review]
Nice addition - I just have a few comments. :)

**Review by MortenTobiasNielsen — CHANGES_REQUESTED** [review]
I added a check to see if the current game is multiplayer and informs the players that it won't work unless it is a multiplayer game.

File: control_info.h [hunk 4]
```diff
@@ -53,6 +53,12 @@ local function debug(msg)
 	end
 end
 
+local function save()
```
[the remark this answers is not in the document]
Comment (theis999) [reply control_info.h:56]: Good

File: Functions.cpp [hunk 5]
```diff
@@ -347,6 +347,11 @@ void launch(std::string task, std::string x_cord, std::string y_cord) {
 	step += 1;
 }
 
+void save(std::string task, std::string amount) {
```
[the remark this answers is not in the document]
Comment (theis999) [reply Functions.cpp:350]: Renamed

File: cMain.cpp [hunk 6]
```diff
@@ -2327,6 +2332,12 @@ void cMain::OnGenerateScript(wxCommandEvent& event) {
 
 		}
 		
+		if (task == "Save") {
+			save(task_number, units);
+			continue;
+
```
[the remark this answers is not in the document]
Comment (theis999) [reply cMain.cpp:2338]: done

File: cMain.cpp (hunk 1 above)
Comment (theis999) [reply cMain.cpp:2944]: Oh that is much easier :)

File: GUI_Base.cpp [hunk 7]
```diff
@@ -585,8 +585,8 @@ GUI_Base::GUI_Base(wxWindow* parent, wxWindowID id, const wxString& title, const
 	wxBoxSizer* bSizer1411;
 	bSizer1411 = new wxBoxSizer(wxVERTICAL);
 
-
-	bSizer1411->Add(100, 0, 1, wxEXPAND, 5);
+	rbtn_save = new wxRadioButton(m_panel1, wxID_ANY, wxT("Save"), wxDefaultPosition, wxDefaultSize, 0);
+	bSizer1411->Add(rbtn_save, 0, wxALL, 5);
```
[the remark this answers is not in the document]
Comment (MortenTobiasNielsen) [reply GUI_Base.cpp:589]: My bad I thought it was a spacer. :)

**Comment by theis999** [comment]
I have updated the pull request with the proposed changes @MortenTobiasNielsen

**Pull request merged**